### PR TITLE
CSC Beam Halo filter update

### PR DESCRIPTION
**Pull request by lathomas**
CSC Beam Halo Filter update. 
-Pairs of segments in opposite endcaps are now matched with a dr/dz condition (rather than dr previously) 
-New flat CSC segment matching added (the old one remains the default for now): unvetoing CSC segments associated to tracker muons with pt < 3 gev 
-Optional dt matching of opposite endcaps segments 
-Testing a CSC segment/calorechit  matching. 
-Two new working points added in BeamHaloSummary.

### CODE REVIEW

**Comment by slava77** [comment]
@cmsbuild please test

@lathomas please post a link to a presentation describing the changes

**Comment by slava77** [comment]
@lathomas please confirm here in this PR thread if you are done with the changes.
We can't follow a moving target.

**Comment by lathomas** [comment]
Hello,
yes I am done !
Sorry for the trouble.

Laurent
Le 4 sept. 2015 19:15, "Slava Krutelyov" <notifications@github.com> a
écrit :

**Comment by lathomas** [comment]
Hi,

The changes on the current filter only affect the counting of the flat segments, especially for opposite endcaps.
I didn't mention it in the PR but the dphi condition has been tightened from 0.35 to 0.2.
For opposite endcaps, the dr<10cm matching condition is now replaced by a dr/dz<0.03 condition, which is looser (it corresponds to dr<  36 cm to 60 cm, depending on the chambers).

The current distribution of the two variables entering the current filter (longest chain of flat segments and opposite endcaps segments found) is shown on S15 here (left plots):
https://indico.cern.ch/event/438533/contribution/1/attachments/1139775/1632303/beamhalo.pdf

These are the only changes to the default filter. 

I have also added various methods to the CSCHaloData class. Two new filters were added to BeamHaloSummary but are not used anywhere so far. So all the changes, except the two mentioned above should not break anything and the size of the updated classes should not be dramatically modified. 

Let me know if you need more information. 

Thanks a lot !

Laurent

File: RecoMET/METAlgorithms/src/CSCHaloAlgo.cc [hunk 1]
```diff
@@ -545,14 +594,67 @@ reco::CSCHaloData CSCHaloAlgo::Calculate(const CSCGeometry& TheCSCGeometry,
...
+         double rhet = rechit.energy()/cosh(rhpos.eta());
+         double dphi_rhseg = fabs(deltaPhi(rhpos.phi(),iGlobalPosition.phi()));
+         double dr_rhseg = sqrt(rhpos.x()*rhpos.x()+rhpos.y()*rhpos.y()) - iR;
+         double dtcorr_rhseg = rechit.time()- fabs(rhpos.z()-iZ)/30- iT; //                                                                       
+         if(rhet>10&&
+            dphi_rhseg <0.05 &&
+            dr_rhseg< 25 && dr_rhseg> -25 && //careful: asymmetric cut might not be the most appropriate thing                                  
+	    dtcorr_rhseg>0
+            ) calomatched= true;
+     }
```
Comment (cvuosalo) [inline RecoMET/METAlgorithms/src/CSCHaloAlgo.cc]: If the three "for" loops in lines 602-644 are intended to be the same, they should be made into a function that is called three times, once each for hbhehits, ecalebhits, and ecaleehits. Maybe there is some subtle difference I'm missing. But copy-and-paste programming should be avoided.

File: RecoMET/METAlgorithms/src/CSCHaloAlgo.cc [hunk 2]
```diff
@@ -561,14 +663,19 @@ reco::CSCHaloData CSCHaloAlgo::Calculate(const CSCGeometry& TheCSCGeometry,
...
 	 float jR =  TMath::Sqrt(jGlobalPosition.x()*jGlobalPosition.x() + jGlobalPosition.y()*jGlobalPosition.y());
-
-	 if (TMath::ACos(TMath::Cos(jPhi - iPhi)) <= max_segment_phi_diff 
-	     && TMath::Abs(jR - iR) <= max_segment_r_diff 
+	 float jZ = jGlobalPosition.z() ;
+	 float jT = jSegment->time();
+	 //	 if(fabs(jZ)<650&& TheEvent.id().run() < 251737)jT-= 25;
+	 if (TMath::ACos(TMath::Cos(jPhi - iPhi)) <= 0.2//max_segment_phi_diff 
+	     //&& TMath::Abs(jR - iR) <= max_segment_r_diff 
+	     && (TMath::Abs(jR - iR) <= max_segment_r_diff || (TMath::Abs(jR - iR)/ TMath::Abs(jZ - iZ)< 0.03 &&  jZ*iZ<0) )
```
Comment (cvuosalo) [inline RecoMET/METAlgorithms/src/CSCHaloAlgo.cc]: Please protect against division by zero if (jZ - iZ) == 0. I don't see any guarantee that such a case would never occur.

File: RecoMET/METAlgorithms/src/CSCHaloAlgo.cc [hunk 3]
```diff
@@ -579,21 +686,37 @@ reco::CSCHaloData CSCHaloAlgo::Calculate(const CSCGeometry& TheCSCGeometry,
...
 	     nSegs++;
 	     minus_endcap = iGlobalPosition.z() < 0 || jGlobalPosition.z() < 0;
 	     plus_endcap = iGlobalPosition.z() > 0 || jGlobalPosition.z() > 0;
+	     //	     if( fabs(jT-iT)/sqrt( (jR-iR)*(jR-iR)+(jZ-iZ)*(jZ-iZ) )<0.05 && fabs(jT-iT)/sqrt( (jR-iR)*(jR-iR)+(jZ-iZ)*(jZ-iZ) )>0.02 && minus_endcap&&plus_endcap ) both_endcaps_dtcut =true;
 	   }
+	   if(Segment1IsGood_alt && Segment2IsGood_alt) {
+             nSegs_alt++;
+             minus_endcap = iGlobalPosition.z() < 0 || jGlobalPosition.z() < 0;
+             plus_endcap = iGlobalPosition.z() > 0 || jGlobalPosition.z() > 0;
+             if( fabs(jT-iT)/sqrt( (jR-iR)*(jR-iR)+(jZ-iZ)*(jZ-iZ) )<0.05 && fabs(jT-iT)/sqrt( (jR-iR)*(jR-iR)+(jZ-iZ)*(jZ-iZ) )>0.02 && minus_endcap&&plus_endcap ) both_endcaps_loose_dtcut_alt =true;
```
Comment (cvuosalo) [inline RecoMET/METAlgorithms/src/CSCHaloAlgo.cc]: Again, please protect against division by zero here. I'm not sure how likely it is for these quantities to be zero, but it is better to be safe.

File: RecoMET/METAlgorithms/src/CSCHaloAlgo.cc [hunk 4]
```diff
@@ -561,14 +663,19 @@ reco::CSCHaloData CSCHaloAlgo::Calculate(const CSCGeometry& TheCSCGeometry,
 	 float jTheta = jGlobalDirection.theta();
 	 float jPhi = jGlobalPosition.phi();
 	 float jR =  TMath::Sqrt(jGlobalPosition.x()*jGlobalPosition.x() + jGlobalPosition.y()*jGlobalPosition.y());
-
-	 if (TMath::ACos(TMath::Cos(jPhi - iPhi)) <= max_segment_phi_diff 
-	     && TMath::Abs(jR - iR) <= max_segment_r_diff 
+	 float jZ = jGlobalPosition.z() ;
+	 float jT = jSegment->time();
+	 //	 if(fabs(jZ)<650&& TheEvent.id().run() < 251737)jT-= 25;
+	 if (TMath::ACos(TMath::Cos(jPhi - iPhi)) <= 0.2//max_segment_phi_diff
```
Comment (slava77) [inline RecoMET/METAlgorithms/src/CSCHaloAlgo.cc]: don't use TMath::, unless required: use std:: functions instead
- replace this acos(cos nonsense with ``` std::abs(reco::deltaPhi ( jPhi - iPhi )); ```

**Comment by lathomas** [comment]
Hello, 
I have implemented your comments. Some of the things reported (like the Acos (cos)) were actually present since a long time but I modified this. 
I protected against zero division. 
I have also written two functions for ECAL and HCAL matching (since the collections to pass are different, there is maybe a better way to do things though...).  
I checked on a beam halo enriched samples that the changes are purely cosmetics (no difference in the filter decision between the two versions). 

I will wait that this PR is approved before updating the 75X one... 
Laurent

**Comment by cvuosalo** [comment]
@cmsbuild please test

**Comment by cvuosalo** [comment]
+1

For #11136 c6f4cd1cf721b613d11821985b1838f5a53c1da1

Small improvements to the CSC beam halo filter.

The code changes are satisfactory, and Jenkins [tests](https://github.com/cms-sw/cmssw/pull/11136) against baseline CMSSW_7_4_X_2015-09-08-2300  show no significant differences. Tests of wokflows 8.0_BeamHalo and 25202.0_TTbar_13 against baseline CMSSW_7_4_11_patch1also show no significant differences. CPU times tests for 25202 show no significant increase in time, with a small addition of 12-17 ms/event for CSC halo processing.

One measure of processing time:
```
	CSCHaloData 	 0.296012 ms/ev -> 12.0602 ms/ev
 Total times: 6478.48 ms/ev -> 6501.05 ms/ev 	 delta: 22.5697
```
Other measures of processing time and memory:
```
Excluding the first event:
  delta/mean delta/orJob     original                   new       module name
  ---------- ------------     --------                  ----       ------------
   +1.949071      +0.02%         0.22 ms/ev ->        17.12 ms/ev CSCHaloData
Job total:  101.357 s/ev ==> 100.159 s/ev

Baseline
Max VSIZ 3606.48 on evt 34 ; max RSS 3044.13 on evt 45
Time av 102.222 s/evt   max 138.68 s on evt 36
M1 Time av 102.24 s/evt   max 138.68 s on evt 36

PR 
Max VSIZ 3606.25 on evt 38 ; max RSS 3018.09 on evt 45
Time av 102.465 s/evt   max 201.66 s on evt 1
M1 Time av 101.027 s/evt   max 137.612 s on evt 18
```

**Comment by lathomas** [comment]
@cvuosalo  @slava77  @davidlange6
I have updated this PR to synchronize it with the recent 76X PR update (now merged): 
https://github.com/cms-sw/cmssw/pull/11528

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by cvuosalo** [comment]
@cmsbuild please test

**Comment by cvuosalo** [comment]
+1

For #11136 f59487aa12fbacbbbbe6fe2b8f67f0fc3809170a

Second approval, after addition of analysis code from #11528. New Jenkins tests still show no significant differences, and a repeat test of workflow 25202.0 against baseline CMSSW_7_4_11_patch1 also shows no significant differences and similar timing results as before (see previous approval comment).

**Comment by lathomas** [comment]
Copy of the following PR for 75X: 
https://github.com/cms-sw/cmssw/pull/11737

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by cvuosalo** [comment]
+1

For #11136 bf4d5c138a70d7caed944777c60bccbc5a02613d

Third approval of this PR for new, small adjustments to the CSC beam halo filter that come from 76X #11737, which has already been merged. #11137 is the matching PR to this one for 75X, and it has already been approved by Reco.

The code changes are satisfactory, and Jenkins tests against baseline CMSSW_7_4_X_2015-10-12-2300 show no significant differences. An additional test of workflow 25202.0_TTbar_13 was done for #11737, and it showed no significant differences.

**Comment by slava77** [comment]
@davidlange6 
I think that this PR can be closed
Even if we do something with 2015, it will be 76X.

**Comment by slava77** [comment]
-1
to be closed

**Pull request closed without merging**